File: python/tests/test_lqr_rrt_star.py
"""LQR-RRT* (Perez et al. 2012): finds a dynamically-feasible, collision-free
path to the goal, fails gracefully when walled off, validates its params, and its
extension heuristics are genuinely LQR-derived (the metric matrix solves the
Riccati equation, and changing Q/R changes the metric and the resulting path)."""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pytest
import yaml
from conftest import CONFIG_DIR, grid_from, open_grid

from navigation.core.params import ParamError, ParamSet
from navigation.global_planning.sampling.lqr_rrt_star import (
    LQRRRTStar,
    lqr_cost_to_go,
    solve_dlqr,
)
from navigation.maps.occupancy_grid import OccupancyGrid2D

_ALGO = "lqr_rrt_star"


def _capped_config(tmp_path: Path, **overrides: float) -> ParamSet:
    """Real config with per-name overrides so tests stay fast and deterministic."""
    doc = yaml.safe_load((CONFIG_DIR / f"{_ALGO}.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] in overrides:
            entry["default"] = overrides[entry["name"]]
    out = tmp_path / f"{_ALGO}.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _motion_valid_path(grid: OccupancyGrid2D, path: list[tuple[float, float]]) -> bool:
    return all(grid.is_motion_valid(a, b) for a, b in zip(path, path[1:], strict=False))


def test_reaches_goal_with_feasible_collision_free_path(tmp_path: Path) -> None:
    grid = open_grid(10, 10, seed=7)
    start, goal = (0.75, 0.75), (9.0, 9.0)
    params = _capped_config(
        tmp_path, max_iterations=1500, goal_bias=0.2, goal_tolerance=1.5, seed=3
    )
    result = LQRRRTStar(params).plan(grid, start, goal)
    assert result.success
    assert result.path[0] == pytest.approx(start)
    # The LQR feedback regulates onto the goal rest-state; the projected endpoint is
    # within tolerance of the goal position.
    assert grid.distance(result.path[-1], goal) <= params.get_float("goal_tolerance")
    # The propagated double-integrator trajectory must be collision-free throughout.
    assert _motion_valid_path(grid, result.path)
    assert result.cost > 0.0


def test_no_path_when_walled_off(tmp_path: Path) -> None:
    # Fully occupied middle column: no trajectory can cross it.
    grid = grid_from(["..#..", "..#..", "..#..", "..#..", "..#.."], seed=3)
    params = _capped_config(tmp_path, max_iterations=300, goal_tolerance=1.0)
    result = LQRRRTStar(params).plan(grid, (0.5, 0.5), (4.5, 4.5))
    assert not result.success
    assert result.path == []
    assert result.cost == 0.0


def test_out_of_range_default_rejected(tmp_path: Path) -> None:
    # r_ctrl default below its declared min must fail validation at load time.
    doc = yaml.safe_load((CONFIG_DIR / f"{_ALGO}.yaml").read_text())
    for entry in doc["params"]:
        if entry["name"] == "r_ctrl":
            entry["default"] = 0.0  # below min 0.001
    bad = tmp_path / f"{_ALGO}.yaml"
    bad.write_text(yaml.safe_dump(doc), encoding="utf-8")
    with pytest.raises(ParamError):
        ParamSet.from_yaml(bad)


def test_metric_matrix_solves_the_riccati_equation() -> None:
    # The distance metric is dist(a,b)=(a−b)ᵀS(a−b) with S the LQR cost-to-go. Prove
    # S is genuinely Riccati-derived (not an ad-hoc form) by checking the discrete
    # algebraic Riccati residual S − (Q + AᵀSA − AᵀSB(R+BᵀSB)⁻¹BᵀSA) ≈ 0.
    q_pos, q_vel, r_ctrl, dt = 1.0, 1.0, 1.0, 0.2
    s, k = solve_dlqr(q_pos, q_vel, r_ctrl, dt)
    a = np.array([[1.0, dt], [0.0, 1.0]])
    b = np.array([[0.5 * dt * dt], [dt]])
    q = np.diag([q_pos, q_vel])
    r = np.array([[r_ctrl]])
    bt_p = b.T @ s
    gain = np.linalg.solve(r + bt_p @ b, bt_p @ a)
    residual = q + a.T @ s @ a - (a.T @ s @ b) @ gain - s
    assert float(np.max(np.abs(residual))) < 1e-9
    # The returned gain is the one induced by S (the Riccati fixed point).
    assert np.allclose(k, gain[0])


def test_lqr_metric_is_a_metric_contract() -> None:
    s, _ = solve_dlqr(1.0, 1.0, 1.0, 0.2)
    at_rest = (1.0, 2.0, 0.0, 0.0)
    # Coincident states cost nothing, the identity of the cost geometry.
    assert lqr_cost_to_go(at_rest, at_rest, s) == 0.0
    # Distinct states cost strictly more than zero.
    for x1 in [(4.0, 6.0, 0.0, 0.0), (1.0, 2.0, 1.0, -1.0), (0.0, 0.0, 0.5, 0.5)]:
        assert lqr_cost_to_go(at_rest, x1, s) > 0.0


def test_control_weight_changes_the_metric() -> None:
    # A different R (control weight) yields a different Riccati S, hence a different
    # cost-to-go for the same maneuver — the metric is derived from the LQR, not fixed.
    a, b = (0.0, 0.0, 0.0, 0.0), (3.0, 2.0, 1.0, -1.0)
    s_cheap, _ = solve_dlqr(1.0, 1.0, 0.2, 0.2)
    s_dear, _ = solve_dlqr(1.0, 1.0, 5.0, 0.2)
    assert lqr_cost_to_go(a, b, s_dear) != pytest.approx(lqr_cost_to_go(a, b, s_cheap))


def test_control_weight_changes_the_path(tmp_path: Path) -> None:
    # End-to-end: a heavier control penalty reshapes the LQR steering, so the planned
    # trajectory differs. Same map/seed/iterations, only r_ctrl changes.
    grid = open_grid(10, 10, seed=5)
    start, goal = (0.75, 0.75), (9.0, 9.0)
    cheap = LQRRRTStar(
        _capped_config(tmp_path, max_iterations=1200, seed=5, r_ctrl=0.2, goal_tolerance=1.5)
    ).plan(grid, start, goal)
    dear = LQRRRTStar(
        _capped_config(tmp_path, max_iterations=1200, seed=5, r_ctrl=50.0, goal_tolerance=1.5)
    ).plan(grid, start, goal)
    assert cheap.success and dear.success
    assert _motion_valid_path(grid, cheap.path)
    assert _motion_valid_path(grid, dear.path)
    # Different LQR gains ⇒ different realised trajectories/costs.
    assert cheap.path != dear.path

File: cpp/tests/test_lqr_rrt_star.cpp
#include <cmath>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/core/params.hpp"
#include "navigation/global_planning/sampling/lqr_rrt_star.hpp"
#include "test_util.hpp"

using namespace navigation;
using core::Point;

namespace {

// Open square grid (resolution 0.5): '.' everywhere.
std::vector<std::string> open_rows(int n) {
  return std::vector<std::string>(static_cast<size_t>(n), std::string(static_cast<size_t>(n), '.'));
}

std::string lqr_cfg(int max_iterations, double goal_tolerance, double r_ctrl = 1.0) {
  return test::write_temp(
      "lqr_rrt_star.yaml",
      "algorithm: lqr_rrt_star\ncategory: global_planning\nparams:\n"
      "  - {name: max_iterations, type: int, default: " + std::to_string(max_iterations) +
          ", min: 1, max: 200000, description: n}\n"
      "  - {name: step_size, type: float, default: 1.5, min: 0.01, max: 100, description: e}\n"
      "  - {name: goal_bias, type: float, default: 0.2, min: 0.0, max: 1.0, description: b}\n"
      "  - {name: goal_tolerance, type: float, default: " + std::to_string(goal_tolerance) +
          ", min: 0.0, max: 100, description: t}\n"
      "  - {name: neighbor_radius, type: float, default: 2.0, min: 0.01, max: 100, description: r}\n"
      "  - {name: q_pos, type: float, default: 1.0, min: 0.001, max: 1000, description: qp}\n"
      "  - {name: q_vel, type: float, default: 1.0, min: 0.001, max: 1000, description: qv}\n"
      "  - {name: r_ctrl, type: float, default: " + std::to_string(r_ctrl) +
          ", min: 0.001, max: 1000, description: rc}\n"
      "  - {name: lqr_dt, type: float, default: 0.2, min: 0.01, max: 10, description: dt}\n"
      "  - {name: control_limit, type: float, default: 10.0, min: 0.01, max: 1000, description: u}\n"
      "  - {name: max_velocity, type: float, default: 1.5, min: 0.01, max: 100, description: v}\n"
      "  - {name: seed, type: int, default: 3, min: 0, max: 2147483647, description: s}\n");
}

}  // namespace

// (a) valid, dynamically-feasible, collision-free path on an open grid --------

TEST(LqrRrtStar, FindsFeasiblePathOnOpenGrid) {
  auto g = test::make_grid(open_rows(20));  // 10m x 10m world
  Point start{0.75, 0.75}, goal{9.0, 9.0};
  global_planning::LqrRrtStarPlanner p(core::ParamSet::from_yaml(lqr_cfg(3000, 1.5)));
  auto r = p.plan(g, start, goal, nullptr);
  ASSERT_TRUE(r.success);
  // The LQR feedback regulates onto the goal rest-state; the projected endpoint is
  // within tolerance of the goal position.
  EXPECT_LE(std::hypot(r.path.back().x - goal.x, r.path.back().y - goal.y), 1.5);
  // The propagated double-integrator trajectory must be collision-free throughout.
  for (size_t i = 0; i + 1 < r.path.size(); ++i) {
    EXPECT_TRUE(g.is_motion_valid(r.path[i], r.path[i + 1])) << "path segment must be collision-free";
  }
  EXPECT_GT(r.cost, 0.0);
}

// (b) no-path case: a full wall separates start from goal --------------------

TEST(LqrRrtStar, NoPathAcrossFullWall) {
  auto rows = open_rows(10);
  rows[5] = "##########";  // impassable horizontal wall
  auto g = test::make_grid(rows);
  Point start{0.25, 0.25}, goal{4.75, 4.75};
  global_planning::LqrRrtStarPlanner p(core::ParamSet::from_yaml(lqr_cfg(400, 1.0)));
  auto r = p.plan(g, start, goal, nullptr);
  EXPECT_FALSE(r.success);
  EXPECT_TRUE(r.path.empty());
}

// (c) param validation failure: out-of-range default -------------------------

TEST(LqrRrtStar, BadControlWeightThrows) {
  std::string bad = test::write_temp(
      "lqr_rrt_star.yaml",
      "algorithm: lqr_rrt_star\ncategory: global_planning\nparams:\n"
      "  - {name: r_ctrl, type: float, default: 0.0, min: 0.001, max: 1000, description: under min}\n");
  EXPECT_THROW(core::ParamSet::from_yaml(bad), std::runtime_error);
}

// (d) the extension heuristics are genuinely LQR-derived ---------------------

// A heavier control weight R yields a different Riccati solution, hence a different
// steering — so the planned trajectory changes. Same map/seed/iterations, only r_ctrl
// differs: proves the metric/steer are derived from the LQR, not fixed (Perez 2012).
TEST(LqrRrtStar, ControlWeightChangesThePath) {
  auto g = test::make_grid(open_rows(20));
  Point start{0.75, 0.75}, goal{9.0, 9.0};
  global_planning::LqrRrtStarPlanner cheap(core::ParamSet::from_yaml(lqr_cfg(2000, 1.5, 0.2)));
  global_planning::LqrRrtStarPlanner dear(core::ParamSet::from_yaml(lqr_cfg(2000, 1.5, 50.0)));
  auto rc = cheap.plan(g, start, goal, nullptr);
  auto rd = dear.plan(g, start, goal, nullptr);
  ASSERT_TRUE(rc.success);
  ASSERT_TRUE(rd.success);
  // Different LQR gains => different realised trajectories. Compare path content
  // (not just length, which can coincide) so the assertion is not fragile.
  bool differ = rc.path.size() != rd.path.size();
  for (size_t i = 0; !differ && i < rc.path.size(); ++i) {
    differ = std::abs(rc.path[i].x - rd.path[i].x) > 1e-9 ||
             std::abs(rc.path[i].y - rd.path[i].y) > 1e-9;
  }
  EXPECT_TRUE(differ);
}
